### PR TITLE
[DataGrid] Add fluid columns width support

### LINKED ISSUE #347: [DataGrid] Columns to have a fluid width
- [x] I have searched the [issues](https://github.com/mui-org/material-ui/issues) of this repository and believe that this is not a duplicate.

## Summary 💡
As the `data-grid` document says, By default, the columns have a width of 100 pixels. We can override it by using the `width` property available in `ColDef`.
Is there an option for columns to have a fluid width so that columns will get a proportion of the width of the container?

## Examples 🌈

```
<div style={{ width: "100%"}}>
<DataGrid
  columns={[
    { field: 'id' },
    { field: 'username' },
    { field: 'age' },
  ]}
  rows={rows}
/>
</div>
```
If my window width is 900px. The column with for 'id', 'username' and 'age' should be 300px each.
Columns should resize automatically when window width changes.

## Motivation 🔦

In most of the table components, I have worked with, this is the default behavior. Is there a way to accomplish this using column props?? or should this be a new feature??

## Benchmark

- https://ag-grid.com/javascript-grid-resizing/#column-flex
- https://www.telerik.com/kendo-react-ui/components/grid/columns/width/
- https://baseweb.design/components/unstable-data-table/

### PR DESCRIPTION
**Pull request by DanailH**
This PR aims to provide support for fluid / responsive columns. This is achieved by setting a `flex` property on a column definition. The `flex` property can work together with the `width` property.

I've tested a couple of data grids to benchmark the functionality and the one that seemed the most intuitive was the one that ag-Grid provides.

The default functionality is not affected by this change. The fluid columns will only work if a user sets the `flex` optional property in the column definition.

Limitations: Currently if you resize a column with flex and you resize the entire page the sizes of all columns are returned to their original state (this is also the current behavior)

PS: Although this solution works if we decide to go with a different public API I will change it.

Fixes #347

### CODE REVIEW

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/columns/columns.md:54]: The usage of "we" is forbidden in the documentation (or at least limited to rare exceptions) https://trello.com/c/1Q7hqXHu/2714-technical-documentation-guidelines

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/columns/columns.md:66]: I think that we can remove these comments. We can include the benchmark in the related issue.

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/columns/columns.md:52]: I think that it should be part of the generic column width concern. 

```suggestion
### Fluid width
```

Comment (oliviertassinari) [reply docs/src/pages/components/data-grid/columns/columns.md:52]: I wonder about 
```suggestion
### Flex width
```

**Review by oliviertassinari — COMMENTED** [review]
@amiranga Does the new API solve your problem? You can preview the documentation at https://deploy-preview-480--material-ui-x.netlify.app/components/data-grid/columns/#column-fluid-width

---

@DanailH for the test side, this feature sounds like a (rare?) case where we could leverage unit tests on the width computation logic rather than using a full end-to-end suite. We could probably have one smoke end to end test in karma (we need to make sure the widths in pixels are correct, screenshots would be bad, and we don't need to have interactions), and cover the edge cases with unit tests? I wonder 🤔

**Comment by oliviertassinari** [comment]
We would need to change the demo or introduce the concept of `minWidth`, the first column is unusable:

IMG_20201021_184427

**Comment by DanailH** [comment]
> We would need to change the demo or introduce the concept of `minWidth`, the first column is unusable:
> 
> IMG_20201021_184427
> 
> (or at least have #416 😁)

Yes, I agree with this and we can easily bump the minWidth. Currently its at `50px` we can make it a bit bigger than that, what do you think @oliviertassinari ?

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (DanailH) [reply docs/src/pages/components/data-grid/columns/columns.md:52]: I was wondering how to structure it but, yes having it as a subsection of the `Column width` does look better.

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (DanailH) [reply docs/src/pages/components/data-grid/columns/columns.md:66]: These links are already in the ticket itself. I saw that we also added some benchmark links in the `resizing` section so that why I added these once but yes without them it looks cleaner ;)

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (DanailH) [reply docs/src/pages/components/data-grid/columns/columns.md:54]: My bad, sorry about that, I'll fix it now.

**Comment by DanailH** [comment]
> @amiranga Does the new API solve your problem? You can preview the documentation at https://deploy-preview-480--material-ui-x.netlify.app/components/data-grid/columns/#column-fluid-width
> 
> @DanailH for the test side, this feature sounds like a (rare?) case where we could leverage unit tests on the width computation logic rather than using a full end-to-end suite. We could probably have one smoke end to end test in karma (we need to make sure the widths in pixels are correct, screenshots would be bad, and we don't need to have interactions), and cover the edge cases with unit tests? I wonder 🤔

Sounds like a plan. The unit tests will check if the width of the columns is calculated correctly after rendering the grid. After all the column calculation when the `flex` is set is a linear formula with 1 unknown.

I'm more concerned about what should we do about the `minWidth`. How it currently works is that I'm taking the sum of the widths of all columns that DOES NOT have `flex` (which by default is 100 px) set and if the sum is more or equal to the total grid width the width of the columns that have `flex` defaults to the base value of 100 px and a scroll may appear (depends on the total size). If on the other hand, the sum of the width of the columns is less than the width of the grid then the available space is distributed among the columns that do have `flex` set. That's why in the end some columns with `flex` may end up with a width of 10px for example. But I can also argue that this is something that can be handled externally.

If we would want to introduce a `minWidth` property on a column configuration I would suggest creating a separate ticket for it because we would also need to handle the case with the `useColumnResize`, what do you thing @oliviertassinari ?

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (oliviertassinari) [reply docs/src/pages/components/data-grid/columns/columns.md]: No problem! I'm doing this too from time to time. Matt is rectifying me when I do it wrong.

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (oliviertassinari) [reply docs/src/pages/components/data-grid/columns/columns.md]: Definitely, I was hesitating about where it would be best to store these benchmarks. What you see in the resizing section is something I forgot to clean up :). I came to the conclusion that the list of a benchmark is living content. it would be too slow to edit them in GIT. It might also create issues if we translate the page.

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (oliviertassinari) [reply docs/src/pages/components/data-grid/columns/columns.md]: I'm raising it because I want on using a flatter structure than ag-grid for the documentation. We will probably in the future need to break it further down as we add more features to the component.

**Comment by oliviertassinari** [comment]
@DanailH Interesting, better ask insights from @dtassone.

File: packages/grid/_modules_/grid/hooks/root/useColumns.ts
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/root/useColumns.ts:52]:
```suggestion
    extendedColumns = extendedColumns.map((column) => {
```

File: packages/grid/_modules_/grid/models/colDef/colDef.ts
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/colDef/colDef.ts:36]:
```suggestion
   * If set, it indicates that a column has fluid width. Range [0, 1].
```

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/models/colDef/colDef.ts:36]: I think that we should match flex-grow API `[0, ∞]` to leverage past learnings from CSS: https://drafts.csswg.org/css-flexbox-1/#propdef-flex-grow
```suggestion
   * If set, it indicates that a column has fluid width. Range [0, ∞].
```

File: packages/grid/_modules_/grid/hooks/root/useColumns.ts
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/root/useColumns.ts:51]:
```suggestion
    const flexMultiplier = containerWidth / flexDivider;
```

**Review by oliviertassinari — COMMENTED** [review]
Unrelated, I have noticed a bug with the hover logic (persistence of hover state on mobile) can be fixed with:

```diff

diff --git a/packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts b/packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts
index fbe307c2..4e0f7bae 100644
--- a/packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts
+++ b/packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts
@@ -126,6 +126,10 @@ export const useStyles = makeStyles(
           cursor: 'col-resize',
           '&:hover, &.Mui-resizing': {
             color: theme.palette.text.primary,
+            // Reset on touch devices, it doesn't add specificity
+            '@media (hover: none)': {
+              color: borderColor,
+            },
           },
         },
         '& .MuiDataGrid-iconSeparator': {
```

File: packages/grid/_modules_/grid/models/colDef/colDef.ts
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/models/colDef/colDef.ts:36]: We also need to update the docs

File: packages/grid/_modules_/grid/hooks/root/useColumns.ts
Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/root/useColumns.ts]: done

**Comment by DanailH** [comment]
> Unrelated, I have noticed a bug with the hover logic (persistence of hover state on mobile) can be fixed with:
> 
> ```diff
> diff --git a/packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts b/packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts
> index fbe307c2..4e0f7bae 100644
> --- a/packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts
> +++ b/packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts
> @@ -126,6 +126,10 @@ export const useStyles = makeStyles(
>            cursor: 'col-resize',
>            '&:hover, &.Mui-resizing': {
>              color: theme.palette.text.primary,
> +            // Reset on touch devices, it doesn't add specificity
> +            '@media (hover: none)': {
> +              color: borderColor,
> +            },
>            },
>          },
>          '& .MuiDataGrid-iconSeparator': {
> ```

Done

**Review by oliviertassinari — APPROVED** [review]

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (dtassone) [inline docs/src/pages/components/data-grid/columns/columns.md:62]: typo:  ...is more than* the width of the grid itself.

**Comment by dtassone** [comment]
Missing storybook stories. Behaviour with many cols, with just a few...
What happens when we resize the grid?

File: packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts
Comment (dtassone) [inline packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts:19]: Should this be selector?

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (DanailH) [reply docs/src/pages/components/data-grid/columns/columns.md:62]: Thanks

File: packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts
Comment (DanailH) [reply packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts:19]: I would keep it as it is. This function is only used in the `hydrateColumns` one and initially, that functionality was part of the `hydrateColumns` but I separated it from it because the `hydrateColumns` was becoming a bit too bloated.

**Comment by DanailH** [comment]
> Missing storybook stories. Behaviour with many cols, with just a few...
> What happens when we resize the grid?
> 
> What happens if we resize the cols?

I'll add examples.

To answer the other 2 questions:
 - If we resize the grid the flex columns will update their size as well (like they are responsive).
 - If we resize a column the width will be updated and a scroll will appear (the flex will no longer be respected). If we resize a column and then resize the entire grid the width from the column resize will not be persisted and the grid will reset.

File: packages/storybook/src/stories/grid-fluid.stories.tsx
Comment (dtassone) [inline packages/storybook/src/stories/grid-fluid.stories.tsx:49]: why not use useDemoData with a transform function

Comment (dtassone) [reply packages/storybook/src/stories/grid-fluid.stories.tsx:49]:
```
export function SmallColSizes() {
  const data = useData(100, 20);
  const transformColSizes = (columns: ColDef[]) => columns.map((c) => ({ ...c, width: 60 }));

  return <XGrid rows={data.rows} columns={transformColSizes(data.columns)} />;
}
```

File: packages/storybook/src/stories/grid-fluid.stories.tsx
Comment (dtassone) [inline packages/storybook/src/stories/grid-fluid.stories.tsx:8]: Could we embed them in `grid-columns.stories.ts` please

File: packages/grid/data-grid/src/DataGrid.test.tsx
Comment (oliviertassinari) [inline packages/grid/data-grid/src/DataGrid.test.tsx:216]: Better use the accessibility features of the grid :)

```suggestion
      const DOMColumns = getAllByRole('columnheader');
```

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (mbrookes) [inline docs/src/pages/components/data-grid/columns/columns.md:54]:
```suggestion
Each column has a fixed width of 100 pixels by default, but column fluidity (responsiveness) can be by achieved by setting the `flex` property in `ColDef`.
```

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (mbrookes) [inline docs/src/pages/components/data-grid/columns/columns.md:56]:
```suggestion
The `flex` property accepts a value between 0 and ∞.
```

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (mbrookes) [inline docs/src/pages/components/data-grid/columns/columns.md:58]:
```suggestion
The `flex` property works by dividing the remaining space in the grid among all flex columns in proportion to their `flex` value. 
For example, consider a grid with a total width of 500px that has three columns: the first with `width: 200`; the second with `flex: 1`; and third with `flex: 0.5`. 
The first column will be 200px wide, leaving 300px remaining. The column with `flex: 1` is twice the size of `flex: 0.5`, which means that final sizes will be: 200px, 200px, 100px.
```

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (mbrookes) [inline docs/src/pages/components/data-grid/columns/columns.md:60]:
```suggestion
Note that `flex` doesn't work together with `width`. If you set both `flex` and `width` in `ColDef`, `flex` will override `width`.
```

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (mbrookes) [inline docs/src/pages/components/data-grid/columns/columns.md:62]:
```suggestion
In addition, `flex` does not work if the combined width of the columns that have `width` is more than the width of the grid itself. If that is the case a scroll bar will be visible and the columns that have `flex` will default back to their base value of 100px.
```

File: packages/grid/data-grid/src/DataGrid.test.tsx
Comment (mbrookes) [inline packages/grid/data-grid/src/DataGrid.test.tsx:261]:
```suggestion
    it('should set the first column to be twice as wide as the second one', () => {
```

**Review by mbrookes — DISMISSED** [review]

File: packages/grid/data-grid/src/DataGrid.test.tsx
Comment (DanailH) [reply packages/grid/data-grid/src/DataGrid.test.tsx]: done

File: packages/storybook/src/stories/grid-fluid.stories.tsx
Comment (DanailH) [reply packages/storybook/src/stories/grid-fluid.stories.tsx]: done

File: packages/storybook/src/stories/grid-fluid.stories.tsx
Comment (DanailH) [reply packages/storybook/src/stories/grid-fluid.stories.tsx]: done

**Review by oliviertassinari — APPROVED** [review]

File: docs/src/pages/components/data-grid/columns/columns.md
Comment (mbrookes) [inline docs/src/pages/components/data-grid/columns/columns.md:64]:
```suggestion
In addition, `flex` does not work if the combined width of the columns that have `width` is more than the width of the grid itself. If that is the case a scroll bar will be visible, and the columns that have `flex` will default back to their base value of 100px.
```

**Comment by DanailH** [comment]
Should I merge this or should I wait for everyone to approve? @mbrookes @dtassone

**Review by dtassone — APPROVED** [review]

**Comment by dtassone** [comment]
> Should I merge this or should I wait for everyone to approve? @mbrookes @dtassone

Thanks for waiting. ATM I would say, wait for the approval of Olivier and me, please :)

File: packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/hooks/features/columns/useColumns.ts:152]: I think that it would be better to keep the same name for variables, `containerWidth` sounds better

**Comment by AleKrabbe** [comment]
Sorry to be asking this here, but it seems related. Is there some sort of solution to make the XGrid table responsive by setting column priorities so that they get hidden based on their priority and width of the screen?

Here is an example: http://www.primefaces.org:8080/showcase/ui/data/datatable/responsive.xhtml?jfwid=2d38a

**Comment by oliviertassinari** [comment]
@AleKrabbe Please open a new issue, it's not related to this effort. Currently, you can implement this with the useMediaQuery helper to match specific breakpoints.

**Pull request merged**
